sw/device/lib/testing/test_coverage.c

```diff
@@ -0,0 +1,84 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/uart.h"
+#include "sw/vendor/llvm_clang_rt_profile/compiler-rt/lib/profile/InstrProfiling.h"
+#include <stdint.h>
+
+/**
+ * When the linker finds a definition of this symbol, it knows to skip loading
+ * the object which contains the profiling runtime’s static initializer.
+ */
+int __llvm_profile_runtime;
+
+/**
+ * Start and end of `.init_array` section where the LLVM profiling runtime
+ * initialization function resides, defined in `flash_link.ld`.
+ */
+extern void (*_init_array_start[])(void);
+extern void (*_init_array_end[])(void);
+
+/**
+ * Computes the CRC32 of a buffer as expected by Python's `zlib.crc32()`. The
+ * implementation below is basically a simplified, i.e. byte-by-byte and without
+ * a lookup table, version of zlib's crc32. See
+ * https://github.com/madler/zlib/blob/2fa463bacfff79181df1a5270fb67cc679a53e71/crc32.c,
+ * lines 111-112 and 276-279.
+ */
+static uint32_t crc32(char *buf, size_t len) {
+  // CRC32 polynomial.
+  static const uint32_t kPoly = 0xEDB88320;
+  // Since we use a contiguous buffer, we don't need to call this function
+  // multiple times. That's why `crc` is not a parameter and is initialized to
+  // zero.
+  uint32_t crc = 0;
+  crc = ~crc;
+  for (; len > 0; --len, ++buf) {
+    crc = crc ^ *buf;
+    for (uint8_t i = 0; i < 8; ++i) {
+      if (crc & 1) {
+        crc = kPoly ^ (crc >> 1);
+      } else {
+        crc = crc >> 1;
+      }
+    }
+  }
```

Comment:
This might be slightly clearer around the control flow:
```c
for (int i  = 0; i < len; ++i) {
  crc ^= buf[i];
  for (int j = 0; j < CHAR_BITS; ++j) {
    crc >>= 1;
    if (crc & 1) { // I might != 0 this personally.
      crc ^= kPoly;
    }
  }
}
```
You don't have to do this but maybe it could help make this bit wrangling a bit clearer.

Reply:
For future reference: This is actually not correct. LSB of `crc` must be checked before it is shifted right, i.e. `crc >>= 1` should not be before the if statement.

Reply:
Oh. You're right. Ugh. Personally I would still write it this way:
```
bool lsb = crc & 1;
crc >>= 1;
if (lsb) {
  crc &= kPoly;
}
```
but *shrug*.